src/compiler/checker.ts

```diff
@@ -14333,16 +14333,25 @@ namespace ts {
         }
 
         // Ordinarily we reduce a keyof M, where M is a mapped type { [P in K as N<P>]: X }, to simply N<K>. This however presumes
-        // that N distributes over union types, i.e. that N<A | B | C> is equivalent to N<A> | N<B> | N<C>. That presumption may not
-        // be true when N is a non-distributive conditional type or an instantiable type with a non-distributive conditional type as
-        // a constituent. In those cases, we cannot reduce keyof M and need to preserve it as is.
-        function maybeNonDistributiveNameType(type: Type | undefined): boolean {
-            return !!(type && (
-                type.flags & TypeFlags.Conditional && (!(type as ConditionalType).root.isDistributive || maybeNonDistributiveNameType((type as ConditionalType).checkType)) ||
-                type.flags & (TypeFlags.UnionOrIntersection | TypeFlags.TemplateLiteral) && some((type as UnionOrIntersectionType | TemplateLiteralType).types, maybeNonDistributiveNameType) ||
-                type.flags & (TypeFlags.Index | TypeFlags.StringMapping) && maybeNonDistributiveNameType((type as IndexType | StringMappingType).type) ||
-                type.flags & TypeFlags.IndexedAccess && maybeNonDistributiveNameType((type as IndexedAccessType).indexType) ||
-                type.flags & TypeFlags.Substitution && maybeNonDistributiveNameType((type as SubstitutionType).substitute)));
+        // that N distributes over union types, i.e. that N<A | B | C> is equivalent to N<A> | N<B> | N<C>. Specifically, we only
+        // want to perform the reduction when the name type of a mapped type is distributive with respect to the type variable
+        // introduced by the 'in' clause of the mapped type. Note that non-generic types are considered to be distributive because
+        // they're the same type regardless of what's being distributed over.
+        function hasDistributiveNameType(mappedType: MappedType) {
+            const typeVariable = getTypeParameterFromMappedType(mappedType);
+            return isDistributive(getNameTypeFromMappedType(mappedType) || typeVariable);
+            function isDistributive(type: Type): boolean {
+                return type.flags & TypeFlags.Conditional ? (type as ConditionalType).root.isDistributive && isDistributiveCheckType((type as ConditionalType).checkType) :
+                    type.flags & (TypeFlags.UnionOrIntersection | TypeFlags.TemplateLiteral) ? every((type as UnionOrIntersectionType | TemplateLiteralType).types, isDistributive) :
+                    type.flags & TypeFlags.IndexedAccess ? isDistributive((type as IndexedAccessType).objectType) && isDistributive((type as IndexedAccessType).indexType) :
+                    type.flags & TypeFlags.Substitution ? isDistributive((type as SubstitutionType).substitute) :
+                    type.flags & TypeFlags.StringMapping ? isDistributive((type as StringMappingType).type) :
+                    true;

```

Comment:
In fact, [another test case](https://www.typescriptlang.org/play?#code/C4TwDgpgBA0hIHsBmBZAhmSATAPAFQBooBVAPigF4oBreZKAbygG1YoBLAOxrqSjyhoAzj0R8AFA2YAFDtxiCReGQF0oEAB7AInLCOJQA-GwBcUThABuEAE4qzwGwFcIAXwCUUe-2Yw1rgG4AWAAoLm0bJDQAY2gURlCoQTMAIwQEABsINE5gkKSUs04nAFsU2zzXUKQnTmjgdgRuJHENMzgxdEwIXBQiNMzszlJPBkSoGwhgJxtuDUrq8QAiFKX3AKA). Again, it should reject the call to f, but fails to do so in the playground, and fails to do so after this fix.

Changing line 14349 to `!(type.flags & TypeFlags.Index)` fixes this particular test case, but if we're trying to be conservative `false` would be better. Or maybe `!!(type.flags & TypeFlags.Primitive)` to avoid being overly conservative

Reply:
Sorry, not sure what's going on with the playground links - they seem to randomly link to other test cases I've been playing with. In case its behaving in a similar way for you, here's the code I'm trying to link to:

```
type KeyofMapped<T, U> = keyof { [ K in keyof T as keyof ({[P in K as T[P] extends U ? K : never]: true}) ]: T[K] };
interface M {
  a: boolean;
  b: number;
}
function f(x: KeyofMapped<M, boolean>) {
  return x;
}
f("b");
```

Reply:
With latest commits now errors as expected.